packages/react-reconciler/src/ReactFiberHooks.js

```diff
@@ -1799,15 +1797,28 @@ function updateSyncExternalStore<T>(
   // Read the current snapshot from the store on every render. This breaks the
   // normal rules of React, and only works because store updates are
   // always synchronous.
-  const nextSnapshot = getSnapshot();
-  if (__DEV__) {
-    if (!didWarnUncachedGetSnapshot) {
-      const cachedSnapshot = getSnapshot();
-      if (!is(nextSnapshot, cachedSnapshot)) {
-        console.error(
-          'The result of getSnapshot should be cached to avoid an infinite loop',
-        );
-        didWarnUncachedGetSnapshot = true;
+  let nextSnapshot;
+  const isHydrating = getIsHydrating();
```

Comment:
I think this was a regression when I made it so that the "rerender" dispatcher is used during the strict mode double render. Originally the idea was to keep the `getIsHydrating` out of the update path because it's always false during an update.

Right now there is no special "rerender" implementation of useSES, but if we add one we can keep this extra check out of the common path.

Reply:
I am skeptical that the double implementations are good for perf (we should benchmark it sometime) but I can change it.

Reply:
You mean because of the code size? Execution wise I don't see why it would be slower

Reply:
Code size and because the function dispatch call is more dynamic so I imagine it can’t inline a jump.

Reply:
Also, where do we stop? Should we have a fourth implementation for the rerender-on-initial-mount case instead of combining it with the rerender-on-update one like we do today? Right now updateEffect has a branch for currentHook === null.

Reply:
Oh I was just thinking the rerender implementation would check `currentHook` and call the mount or update ones as appropriate. I agree it's not worth duplicating the entire update implementation.

Like I did here for rerenderOptimisticHook:

https://github.com/facebook/react/blob/df12d7eac40c87bd5fdde0aa5a739bce9e7dce27/packages/react-reconciler/src/ReactFiberHooks.js#L2049-L2052

The main reason we added the separate phases was to keep the _mount_ path fast, not so much the update one. But subjectively I also prefer it for code organization reasons because the logic does end up being significantly different in a lot of the hooks.

Reply:
Ultimately though my original comment was only a factoring nit. I'd probably structure this differently if I were doing it myself, but I liked how you had it before more than this way, so you could just revert it back to that and if I really want to I can submit a follow up PR later.

The important thing is the fix itself, and the regression test you added.

Reply:
Happy to change it to what you just described! That way makes sense to me. Wasn’t sure if you approved of the fix originally.

Reply:
Hmm I'm actually not sure the best way to reuse the mount code on rerender because we need it to call updateWorkInProgressHook including for the interior mountEffect. Gonna land it with the unified update+rerender code like I had before but would be happy to review a refactor if you have one in mind.